hist/histv7/inc/ROOT/RHistAutoAxisFiller.hxx

```diff
@@ -0,0 +1,147 @@
+Feedback is welcome!
+*/
+template <typename BinContentType>
+class RHistAutoAxisFiller final {
+   /// The filled histogram, after it has been constructed
+   std::optional<RHist<BinContentType>> fHist;
+
+   /// The number of normal bins
+   std::size_t fNNormalBins;
+   /// The maximum buffer size until Flush() is automatically called
+   std::size_t fMaxBufferSize;
+
+   /// The buffer of filled entries
+   // FIXME: need to store weights!
+   std::vector<double> fBuffer;
+   /// The minimum of the filled entries
+   double fMinimum = std::numeric_limits<double>::infinity();
+   /// The maximum of the filled entries
+   double fMaximum = -std::numeric_limits<double>::infinity();
+
+public:
+   /// Create a filler object.
+   ///
+   /// \param[in] nNormalBins the number of normal bins, must be > 0
+   /// \param[in] maxBufferSize the maximum buffer size, must be > 0
+   explicit RHistAutoAxisFiller(std::size_t nNormalBins, std::size_t maxBufferSize = 1024)
+      : fNNormalBins(nNormalBins), fMaxBufferSize(maxBufferSize)
+   {
+      if (nNormalBins == 0) {
+         throw std::invalid_argument("nNormalBins must be > 0");
+      }
+      if (maxBufferSize == 0) {
+         throw std::invalid_argument("maxBufferSize must be > 0");
+      }
+   }
+
+   std::size_t GetNNormalBins() const { return fNNormalBins; }
+   std::size_t GetMaxBufferSize() const { return fMaxBufferSize; }
+
+   /// Fill an entry into the histogram.
+   ///
+   /// \param[in] x the argument
+   void Fill(double x)
+   {
+      // If the histogram exists, forward the Fill call.
+      if (fHist) {
+         fHist->Fill(x);
+         return;
+      }
+
+      fBuffer.push_back(x);
+      fMinimum = std::min(fMinimum, x);
+      fMaximum = std::max(fMaximum, x);
+
+      if (fBuffer.size() >= fMaxBufferSize) {
+         Flush();
+      }
+   }
+
+   /// Flush the buffer of entries and construct the histogram.
+   ///
+   /// Throws an exception if the buffer is empty, the axis interval cannot be determined, or if it would be empty
+   /// because the minimum equals the maximum.
+   void Flush()
+   {
+      if (fHist) {
+         assert(fBuffer.empty() && "buffer should have been emptied");
+         return;
+      }
+
+      if (fBuffer.empty()) {
+         throw std::runtime_error("buffer is empty, cannot create histogram");
+      }
+      if (!std::isfinite(fMinimum) || !std::isfinite(fMaximum)) {
+         throw std::runtime_error("could not determine axis interval");
+      }
+      if (fMinimum == fMaximum) {
+         throw std::runtime_error("axis interval is empty");
+      }
+
+      // Slightly increase the upper limit to make sure the maximum is included in the last bin.
+      double high = std::nextafter(fMaximum, std::numeric_limits<double>::infinity());
+      assert(high > fMaximum);
+      fHist.emplace(fNNormalBins, std::make_pair(fMinimum, high));
```

Comment:
I would suggest to extend the range with a margin of a few percent to each side.

I have anecdotal evidence that many people do it anyway when they create plots because it looks a bit more pleasing, but it also ensures that future values close to the min/max will be inside the range. The threshold is obviously arbitrary, but a margin of e.g. 5% could work. Note also that this will probably be used for the first attempts, so it's more important the the entire distribution is in range than computing the tightest possible range.
```
const auto margin = 0.05 * (fMaximum - fMinimum);
const auto high = fMaximum + margin;
const auto low = fMinimum - margin;
```